analyzers/src/SonarAnalyzer.Common/Helpers/SonarAnalysisContext.cs

```diff
@@ -92,6 +92,35 @@ internal SonarAnalysisContext(AnalysisContext context, IEnumerable<DiagnosticDes
         public bool ShouldAnalyzeGenerated(Compilation c, AnalyzerOptions options) =>
             ShouldAnalyzeGenerated(context, c, options);
 
+        public bool IsTestProject(Compilation c, AnalyzerOptions options) =>
+            IsTestProject(context.TryGetValue, c, options);
+
+        // FIXME: replace with caching at higher level
```

Comment:
@pavel-mikula-sonarsource  - this is only used when reporting issues (invoked in `AnalysisContextExtensions`), and we have a problem.

SyntaxNodeAnalysisContext, SyntaxTreeAnalysisContext, SymbolAnalysisContext, CodeBlockAnalysisContext don't have the `TryGetValue` in them. So we cannot use the cache. 

And this makes the current process _extremely_ inefficient, because we read the config file each time we report via SyntaxNodeAnalysisContext, SyntaxTreeAnalysisContext, SymbolAnalysisContext, CodeBlockAnalysisContext.

I think the best way would be to read the file in SonarDiagnosticAnalyzer.Initialize, but we are missing the `AnalyzerOptions` at that point. 

The other option is a bit messy: we could do this computation in each rule ... in the registration callback... and get whether it's a test project from the cache at that point, and use it later when reporting issues... but this means changing _every_single_rule.

Reply:
Let's do baby steps. Keep it as it is for now. The XML file is intentionally very small => not so expensive to load.

I'll try to find a caching mechanism meanwhile. I'll investigate Roslyn APIs first, but some kid of file-path based logic would do the trick as well.

Reply:
Ok.

I thought about it in the past few days and one option is to drop this check from reporting. From our previous discussion, this is only done for the case where one class implements multiple rules, and one rule may be TEST-scope, and another MAIN-scope. And the class gets registered, but when reporting issues, we need to do another check.

Doing this check for _ALL_ rules (most of which are in a single class) seems a waste of logic and an unnecessary complexity. This reporting-level extra check can be done just by the classes which implement multiple rules (and it's only a handful of them), and the other ones can use reporting without checking the project type.

Reply:
This idea LGTM.